File: go/raindrops/raindrops.go
// Package raindrops makes the sound of raindrops for some numbers
package raindrops

import (
	"strconv"
	"strings"
)

const testVersion = 3

// Convert transforms a number into raindrop sounds..or a number string
func Convert(x int) string {
	d3 := divisibleBy(x, 3)
	d5 := divisibleBy(x, 5)
	d7 := divisibleBy(x, 7)
	if !(d3 || d5 || d7) {
		return strconv.Itoa(x)
	}
	drops := make([]string, 0)
	if d3 {
		drops = append(drops, "Pling")
	}
	if d5 {
		drops = append(drops, "Plang")
	}
	if d7 {
		drops = append(drops, "Plong")
	}
	return strings.Join(drops, "")
}

func divisibleBy(x, y int) bool {
	return x%y == 0
}

File: ruby/raindrops/raindrops.rb
# frozen_string_literal: true

# Create raindrop sounds for a number based on divisors
class Raindrops
  NUM_SOUND_PAIRS = {
    3 => 'Pling',
    5 => 'Plang',
    7 => 'Plong',
  }.freeze

  def self.convert(num)
    sounds = NUM_SOUND_PAIRS
             .select { |key| divisible?(num, key) }
             .values
             .join
    sounds.empty? ? num.to_s : sounds
  end

  def self.divisible?(num, div)
    (num % div).zero?
  end
end
